File: frontend/src/api/workflows.ts
import client from './client'
import type { WorkflowRun } from '@/types/models'

export interface WorkflowArtifactsResponse {
  ok: boolean
  run: WorkflowRun
  artifacts: Record<string, { label: string; path: string; content: string }>
}

export const workflowsAPI = {
  list(limit = 20) {
    return client.get<any, { ok: boolean; runs: WorkflowRun[] }>(`/workflows/runs?limit=${limit}`)
  },

  getCurrent(siteId: string) {
    return client.get<any, { ok: boolean; run: WorkflowRun | null }>(`/workflows/sites/${siteId}/current`)
  },

  create(siteId: string, payload: { name?: string; enabled_mcp_services?: string[]; enabled_skill_ids?: string[] }) {
    return client.post<any, { ok: boolean; run: WorkflowRun }>(`/workflows/sites/${siteId}/runs`, payload)
  },

  get(runId: string) {
    return client.get<any, { ok: boolean; run: WorkflowRun }>(`/workflows/runs/${runId}`)
  },

  generateStage(runId: string, payload: { stage?: string; content?: string; notes?: string }) {
    return client.post<any, { ok: boolean; run: WorkflowRun; stage: string; content: string; artifact_path: string }>(
      `/workflows/runs/${runId}/generate-stage`,
      payload,
    )
  },

  confirmStage(runId: string) {
    return client.post<any, { ok: boolean; run: WorkflowRun }>(`/workflows/runs/${runId}/confirm-stage`)
  },

  getArtifacts(runId: string) {
    return client.get<any, WorkflowArtifactsResponse>(`/workflows/runs/${runId}/artifacts`)
  },
}

File: backend/api/v2/workflows.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Body, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from backend.api.deps import get_current_user, get_db
from backend.services.workflow_service import workflow_service

router = APIRouter(prefix="/workflows")


@router.get("/runs")
async def list_workflow_runs(
    limit: int = Query(default=20, ge=1, le=100),
    current_user: object = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    return {"ok": True, "runs": await workflow_service.list_runs(db, current_user, limit=limit)}


@router.get("/sites/{site_id}/current")
async def get_current_workflow_run(
    site_id: str,
    current_user: object = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    return {"ok": True, "run": await workflow_service.get_current_run(db, current_user, site_id)}


@router.post("/sites/{site_id}/runs")
async def create_workflow_run(
    site_id: str,
    payload: dict[str, Any] = Body(default_factory=dict),
    current_user: object = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    return {"ok": True, "run": await workflow_service.create_run(db, current_user, site_id, payload)}


@router.get("/runs/{run_id}")
async def get_workflow_run(
    run_id: str,
    current_user: object = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    return {"ok": True, "run": await workflow_service.get_run_detail(db, current_user, run_id)}


@router.post("/runs/{run_id}/generate-stage")
async def generate_workflow_stage(
    run_id: str,
    payload: dict[str, Any] = Body(default_factory=dict),
    current_user: object = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    return await workflow_service.generate_stage(db, current_user, run_id, payload)


@router.post("/runs/{run_id}/confirm-stage")
async def confirm_workflow_stage(
    run_id: str,
    current_user: object = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    return await workflow_service.confirm_stage(db, current_user, run_id)


@router.get("/runs/{run_id}/artifacts")
async def get_workflow_artifacts(
    run_id: str,
    current_user: object = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> dict[str, Any]:
    return await workflow_service.get_artifacts(db, current_user, run_id)
